pkg/controller/operconfig/migration.go

```diff
@@ -97,3 +132,388 @@ func convertEgressFirewallToEgressNetworkPolicy(ctx context.Context, client cnoc
+		nodeObj, err := client.Default().Kubernetes().CoreV1().Nodes().Get(ctx, hostStr, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if nodeObj.Labels == nil {
+			nodeObj.Labels = make(map[string]string)
+		}
+		if _, ok := nodeObj.Labels[egressAssignable]; !ok {
+			nodeObj.Labels[egressAssignable] = ""
+		}
+
+		// copy egressCIDRs to node label if possible
+		// - if egressCIDRs contains values (automatic config), on rollback this annotation will provide the respective hostsubnet with its original egressCIDRs values
+		// - if egressCIDRs is empty (manual config), on rollback this label will not exist, and so we default to node's subnet for egressCIDR field.
+		egressCIDRs, found, err := unstructured.NestedStringSlice(hsn.Object, "egressCIDRs")
+		if err != nil {
+			return fmt.Errorf("egressCIDRs not found in HostSubnet object %s, probable underlying error: %v", hsn.GetName(), err)
+		} else if found {
+			// automatic configuration
+			ovnMigrationAnnotation := OVNMigrationNodeAnnotation{
+				EgressCIDRs: egressCIDRs,
+			}
+
+			if nodeObj.Annotations == nil {
+				nodeObj.Annotations = make(map[string]string)
+			}
+			if _, ok := nodeObj.Annotations[egressCIDRAnnotationName]; !ok {
+				egressCIDRsText, err := json.Marshal(ovnMigrationAnnotation)
+				if err != nil {
+					return err
+				}
+				nodeObj.Annotations[egressCIDRAnnotationName] = string(egressCIDRsText)
+			}
+		}
+		_, err = client.Default().Kubernetes().CoreV1().Nodes().Update(ctx, nodeObj, metav1.UpdateOptions{})
+		return err
+	}); err != nil {
+		return err
+	}
+
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		hsn, err := client.Default().Dynamic().Resource(gvrHostSubnet).Get(ctx, hsn.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		// now remove egressIP config from hostsubnet
+		hsn.Object["egressCIDRs"] = nil
+		hsn.Object["egressIPs"] = nil
+		_, err = client.Default().Dynamic().Resource(gvrHostSubnet).Update(ctx, hsn, metav1.UpdateOptions{})
+		return err
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func deleteCloudPrivateIpConfigs(ctx context.Context, client cnoclient.Client, egressIps []interface{}) error {
+	for _, egressIp := range egressIps {
+		egressIpStr := fmt.Sprintf("%v", egressIp)
+		// delete cloudprivateipconfig
+		err := client.Default().Dynamic().Resource(gvrCloudPrivateIPConfig).Delete(ctx, egressIpStr, metav1.DeleteOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				continue
+			} else {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func applyEgressIpList(ctx context.Context, client cnoclient.Client, egressIpList []*uns.Unstructured) error {
+	for _, egressIp := range egressIpList {
+		if err := apply.ApplyObject(ctx, client, egressIp, ""); err != nil {
+			return err
+		}
+	}
+
+	netNamespaceList, err := cnoclient.ListAllOfSpecifiedType(gvrNetnamespace, ctx, client)
+	if err != nil {
+		return err
+	}
+
+	for _, nns := range netNamespaceList {
```

Comment:
To answer this, we should look at the overall function being split as a whole 
```
case string(operv1.NetworkTypeOVNKubernetes):
	egressIpList, err := convertSdnEgressIpToOvnEgressIp(ctx, client)
	if err != nil {
		return err
	}
	return applyEgressIpList(ctx, client, egressIpList)
```

The function is split here because we can return the egressIpList prior to it being applied. This is useful for my unit test, and doesn't affect the code since the egressIpList gets applied in the function afterward anyway. 

This is needed for the SDNtoOVN function's unit test because fakeClient [has issues with ApplyObject function](https://github.com/kubernetes/kubernetes/issues/104052) Meanwhile, my OVNtoSDN function does not use ApplyObject and so fakeClient has no issues.

So yes we can move the CloudPrivateIpConfig deletion into the second loop, but we cannot consolidate everything into one loop/one overall function. I think this can remain the same unless there is some issue with functionality.

Reply:
Perhaps we could pass the list of netns here as well, to make sure we are acting on the same ones that we originally saw.

Reply:
I don't think this is necessary. The netnamespaces are created automatically by SDN, for every namespace that exists. Due to the readiness check in my code (see below) 
```egressIpRollbackReady, err := sdnEgressIpResourcesReady(ctx, client)```
...the cluster will have all the netnamespaces matching the namespaces it needs for rollback migration.

Once this condition is met, the CNO has stabilized and we don't expect a lot of restarts/etc. that could cause a netnamespace or namespace to disappear, and the CNO code itself does not delete namespaces or netnamespaces at any point. The only edge case is if a user deletes these resources, which...they shouldn't be configuring their cluster during a rollback at all, so I don't think we need to concern ourselves with this.

Reply:
I am not talking about the rollback.

In convertSDNtoOVN we read the NetNS and build a list of egress IPs. 
The we pass the list of egress ips to applyEgressIPs where we read again the NetNS.

What I mean is returning back from convertSDNtoOVN the list of NetNS as well and passing it to applyEgressIPs along with the egressIPs.

I understand how improbable it is that the list of NetNS read in convertSDNtoOVN is different from the one read in applyEgressIPs, but the way I see it there is no reason to risk it if the solution is so trivial, avoids an extra read and it just feels more consistent all around.

I will leave it up to you though.

Reply:
Sure, I can make this change, it does avoid an extra read.

(Separate reminder for myself - unstructured is imported twice - fix this.)